docs/docs-beta/docs/guides/external-systems/non-python.md

```diff
@@ -1,4 +1,141 @@
+    return {
+      xs: tf.tensor2d(Object.values(xs), [Object.values(xs).length, 1]),
+      ys: tf.tensor2d(Object.values(ys), [Object.values(ys).length, 1])
+    };
+  })
+  .batch(100);
+
+  const model = tf.sequential()
+  model.add(tf.layers.dense({units: 1, inputShape: [1]}));
+  model.compile({loss: 'meanSquaredError', optimizer: 'sgd'});
+
+  await model.fitDataset(dataset, {epochs: 250})
+  await model.save(path_to_model);
+  model.summary();
+  return model;
+}
+```
+
+# Define your asset in Dagster
+
+<CodeExample filePath="guides/automation/pipes-asset.py" language="python" title="Asset using Dagster Pipes." />
+
+Dagster Pipes follows a similar design to Unix pipes, hence the name. The `PipesSubprocessClient` is responsible for running external processes and setting up input/output files. The asset defined here is materialized using the `PipesSubprocessClient` running a Node.js file containing the `train_model` function.
+
+# Add JavaScript utility functions to access pipes
+
+The `PipesSubprocessClient` calls the child process with two environment variables defined, each containing a path to a file. One for input and one for output.
+- DAGSTER_PIPES_CONTEXT: Input context
+- DAGSTER_PIPES_MESSAGES: Output context
+
+Both environment variables are base64, zip compressed JSON objects with a "path" key. These functions decode these environment variables and access the files to hook up our JavaScript function to Dagster.
+
+```javascript
+import * as util from 'util';
+import { promises as fs } from "fs";
+import { inflate } from 'zlib';
+
+const inflateAsync = util.promisify(inflate);
+
+async function _decodeParam(value) {
+  if (!value) {
+    return null;
+  }
+  // Decode from base64 and unzip
+  const decoded = Buffer.from(value, "base64");
+  const decompressed = await inflateAsync(decoded);
+  // Deserialize to JSON
+  return JSON.parse(decompressed.toString("utf-8"));
+}
+
+async function getPipesContext() {
+  // get the env var value for where the pipes context is stored
+  const encodedPipesContextParam = process.env.DAGSTER_PIPES_CONTEXT;
+  // decove the value to get the input file path
+  const decodedPipesContextParam = await _decodeParam(encodedPipesContextParam);
+  if (!decodedPipesContextParam) {
+    return null;
+  }
+  return await fs.readFile(decodedPipesContextParam.path, "utf-8")
+    .then((data) => JSON.parse(data));
+}
+
+async function setPipesMessages(message) {
+  // get the env var value for where the pipes message is stored
+  const encodedPipesMessagesParam = process.env.DAGSTER_PIPES_MESSAGES;
+  // decode the value to get the output file path
+  const decodedPipesMessagesParam = await _decodeParam(encodedPipesMessagesParam);
+  if (!decodedPipesMessagesParam) {
+    return null;
+  }
+  const path = decodedPipesMessagesParam.path;
+  await fs.appendFile(path, JSON.stringify(message) + "\n");
+}
+```
+
+# Create an API for Dagster to invoke
```

Comment:
Part of my confusion may be the overloaded "API" term, and in the JavaScript world my first thought was HTTP based API, that's my b.

Alternatively we could name this an "entrypoint" for Dagster?

Reply:
Yeah makes sense, went with "interface with" but could change it to entrypoint if you feel strongly.